eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,236 @@
+     */
+    private Map<DetailAST, String> getInvalidTypeFromTypeArguments(DetailAST typeToken) {
+        final Map<DetailAST, String> invalidTypesMap = new HashMap<>();
+
+        if (typeToken == null) {
+            return invalidTypesMap;
+        }
+
+        final DetailAST identToken = typeToken.findFirstToken(TokenTypes.IDENT);
+        // if there is no IDENT token, implies the token is a default java type.
+        if (identToken == null) {
+            return invalidTypesMap;
+        }
+
+        final String typeName = identToken.getText();
+        // if does not exist in the classPathMap, that implies the type is java default types, such as int.
+        if (classPathMap.containsKey(typeName)) {
+            if (!isValidLibrary(classPathMap.get(typeName))) {
+                invalidTypesMap.put(typeToken, typeName);
+            } else {
+                // Checks multiple type arguments
+                final DetailAST typeArgumentsToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENTS);
+                if (typeArgumentsToken != null) {
+                    for (DetailAST ast = typeArgumentsToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+                        if (ast.getType() == TokenTypes.TYPE_ARGUMENT) {
+                            String invalidTypeName = getInvalidTypeNameFromTypeArgument(ast);
+                            if (invalidTypeName != null) {
+                                invalidTypesMap.put(ast, invalidTypeName);
+                            }
+                        }
+                    }
+                } else {
+                    // Checks single type argument
+                    final DetailAST typeArgumentToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENT);
+                    String invalidTypeName = getInvalidTypeNameFromTypeArgument(typeArgumentToken);
+                    if (invalidTypeName != null) {
+                        invalidTypesMap.put(typeArgumentToken, invalidTypeName);
+                    }
+                }
+            }
+        }
+        return invalidTypesMap;
+    }
+
+    /**
+     * Get invalid type name.
+     *
+     * @param typeArgumentToken TYPE_ARGUMENT AST node
+     * @return an invalid type name if it is an invalid library. Otherwise, returns null.
+     */
+    private String getInvalidTypeNameFromTypeArgument(DetailAST typeArgumentToken) {
+        if (typeArgumentToken == null) {
+            return null;
+        }
+
+        final DetailAST identToken = typeArgumentToken.findFirstToken(TokenTypes.IDENT);
+        // if there is no IDENT token, implies the token is default java types.
+        if (identToken != null) {
+            final String typeName = identToken.getText();
+            // if not exist in the classPathMap, that implies the type is java default types, such as int.
+            if (classPathMap.containsKey(typeName) && !isValidLibrary(classPathMap.get(typeName))) {
```

Comment:
This first check can be rolled into your isValidLibrary method because it appears you're also doing a validity check in this if statement.

```java
private String getInvalidTypeNameFromTypeArgument(DetailAST typeArgumentToken) {
…
  if (identToken == null) {
     return null;
  }

  final String typeName = identToken.getText();
  return isValidLibrary(typeName) ? null : typeName;
}

private boolean isValidLibrary(String typeName) {
   // If the qualified class name does not exist in the map, it implies the type is a primitive Java type (ie. int, long, etc).
   if (!classPathMap.containsKey(typeName)) {
      return true;
   }

   String qualifiedName = classPathMap.get(typeName);

   return VALID_DEPENDENCY_PACKAGE_NAMES.stream()
         .anyMatch(validPackageName -> qualifiedName.startsWith(validPackageName));
}
```

Reply:
You're not checking if the "library" is valid, you're checking if it's a valid class dependency. imho, a library is a collection of related classes